Fix issue #20 : add getAvailableCurrencies method for ProviderInterface

slavielle opened the pull request.

ojhaujjwal commented on src/Provider/ProviderInterface.php:
```diff
@@ -11,4 +12,11 @@
+     *
+     * @return array
+     */
+    public function getSupportedCurrencies();
```
how about creating a different interface like `ProvidesSupportedCurrencies` for this?

slavielle replied:
Hello ojhaujjwal, 

I would say it's a bit of over designing but you're the captain on board captain ;)

I'll do this shortly

slavielle replied:
Well thinking about it, putting `getSupportedCurrencies` in another interface may cause an inconsistency back to the `CurrencyConverter` class ... 

let's consider this :
```php
    public function setCacheAdapter(Cache\Adapter\CacheAdapterInterface $cacheAdapter)
    {
        $this->setCachable(true);
        $this->cacheAdapter = $cacheAdapter;

        return $this;
    } 
```

if we check `$cacheAdapter` is `Cache\Adapter\CacheAdapterInterface` here, then we can't be sure `getSupportedCurrencies` is implemented in $cacheAdapter if declaration is not included in CacheAdapterInterface ... see ?

ojhaujjwal commented:
we also need to check if the currencies are supported in the main class while calculating the exchange rate.

slavielle commented:
> we also need to check if the currencies are supported in the main class while calculating the exchange rate.

are you talking about `CurrencyConverter::convert` method ?

slavielle commented:
Thinking about it, supported currencies also require caching ... I implemented it in the former commit.
I also implemented checking if the currencies are supported in the main class while calculating the exchange rate as you asked for.

What do you think about this last commit ?

ojhaujjwal commented:
+1 man. I like the overall idea. I will weigh in my comments.

ojhaujjwal commented on src/Provider/FixerApi.php:
```diff
@@ -45,9 +45,18 @@ public function getRate($fromCurrency, $toCurrency)
+     */
+    public function getSupportedCurrencies()
+    {
+        $result = json_decode($this->httpClient->get(self::FIXER_API_BASEPATH)->getBody(), true);
```
I get forbidden at times while accessing this endpoint.

ojhaujjwal commented on src/Cache/Adapter/FileSystem.php:
```diff
@@ -99,4 +118,37 @@ public function getRate($fromCurrency, $toCurrency)
+    /**
+     * {@inheritDoc}
+     */
+    function supportedCurrenciesCacheExists()
```
please don't omit method visibility

ojhaujjwal commented on src/Cache/Adapter/ProvidesSupportedCurrenciesCacheInterface.php:
```diff
@@ -0,0 +1,29 @@
+
+interface ProvidesSupportedCurrenciesCacheInterface
+{
+    const   SUPPORTED_CURRENCIES_CACHE_KEY_NAME = 'supported-currencies';
```
unnecessary space

ojhaujjwal commented:
Apologies for delays. I will write all my comments this weekend.
